File: Java/src/com/company/Trie/LC208_Trie.java
package com.company.Trie;

public class LC208_Trie {
    private class TrieNode{
        public TrieNode(){
            children = new TrieNode[26];
            is_word = false;
        }
        public TrieNode[] children;
        public boolean is_word;
    }
    /** Initialize your data structure here. */
    private TrieNode root;

    public LC208_Trie() {
        root = new TrieNode();
    }

    /** Inserts a word into the trie. */
    public void insert(String word) {

        TrieNode cur_node = root;
        char[] wArr = word.toCharArray();
        for(char c : wArr){
            if(cur_node.children[c - 'a'] == null){
                cur_node.children[c - 'a'] = new TrieNode();
            }
            cur_node = cur_node.children[c - 'a'];
        }
        cur_node.is_word = true;

    }

    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode cur_node = root;
        char[] wArr = word.toCharArray();
        for(char c : wArr){
            if(cur_node.children[c - 'a'] == null) return false;
            cur_node = cur_node.children[c - 'a'];
        }
        return cur_node.is_word;

    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode cur_node = root;
        char[] pArr = prefix.toCharArray();
        for(char c : pArr){
            if(cur_node.children[c - 'a'] == null) return false;
            cur_node = cur_node.children[c - 'a'];
        }
        return true;

    }
}

File: python/Trie/lc208_Trie.py
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = {}

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        cur_node = self.root
        for c in word:
            if c not in cur_node:
                cur_node[c] = {}
            cur_node = cur_node[c]
        cur_node['#'] = 1
        return

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        cur_node = self.root
        for c in word:
            if c not in cur_node:
                return False
            cur_node = cur_node[c]

        if '#' in cur_node:
            return True

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        cur_node = self.root
        for c in prefix:
            if c not in cur_node:
                return False
            cur_node = cur_node[c]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
